python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,355 @@
+                send_timestamp,
+                window_avg,
+            )
+
+    def record_request_metrics_for_handle(
+        self,
+        *,
+        handle_id: str,
+        actor_id: Optional[str],
+        handle_source: DeploymentHandleSource,
+        queued_requests: float,
+        running_requests: Dict[ReplicaID, float],
+        send_timestamp: float,
+    ) -> None:
+        """Records average number of queued and running requests at a handle for this
+        deployment.
+        """
+
+        if (
+            handle_id not in self._handle_requests
+            or send_timestamp > self._handle_requests[handle_id].timestamp
+        ):
+            self._handle_requests[handle_id] = HandleRequestMetric(
+                actor_id=actor_id,
+                handle_source=handle_source,
+                queued_requests=queued_requests,
+                running_requests=running_requests,
+                timestamp=send_timestamp,
+            )
+
+    def drop_stale_handle_metrics(self, alive_serve_actor_ids: Set[str]) -> None:
+        """Drops handle metrics that are no longer valid.
+
+        This includes handles that live on Serve Proxy or replica actors
+        that have died AND handles from which the controller hasn't
+        received an update for too long.
+        """
+
+        timeout_s = max(
+            2 * self.autoscaling_policy_manager.get_metrics_interval_s(),
+            RAY_SERVE_MIN_HANDLE_METRICS_TIMEOUT_S,
+        )
+        for handle_id, handle_metric in list(self._handle_requests.items()):
+            # Drop metrics for handles that are on Serve proxy/replica
+            # actors that have died
+            if (
+                handle_metric.is_serve_component_source
+                and handle_metric.actor_id not in alive_serve_actor_ids
+            ):
+                del self._handle_requests[handle_id]
+                if handle_metric.total_requests > 0:
+                    logger.debug(
+                        f"Dropping metrics for handle '{handle_id}' because the Serve "
+                        f"actor it was on ({handle_metric.actor_id}) is no longer "
+                        f"alive. It had {handle_metric.total_requests} ongoing requests"
+                    )
+            # Drop metrics for handles that haven't sent an update in a while.
+            # This is expected behavior for handles that were on replicas or
+            # proxies that have been shut down.
+            elif time.time() - handle_metric.timestamp >= timeout_s:
+                del self._handle_requests[handle_id]
+                if handle_metric.total_requests > 0:
+                    logger.info(
+                        f"Dropping stale metrics for handle '{handle_id}' "
+                        f"because no update was received for {timeout_s:.1f}s. "
+                        f"Ongoing requests was: {handle_metric.total_requests}."
+                    )
```

Comment:
Is it simple to get the `actor_id` here? It would be useful to include that in this log for debugging.

Reply:
Good suggestion, thanks! I've added actor id info to this log.